<a href="https://colab.research.google.com/github/phqnrz/machine-learning/blob/main/1_intro.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## __1. Introdução ao Aprendizado de Máquina:__
<font size=3>

- __1.1__ Qual a relação do Aprendizado de Máquina (AM) com a área da Inteligência Artificial (IA)?
- __1.2__ O que significa _"aprendizado"_ em AM?
- __1.3__ O que podemos fazer com o AM?
- __1.4__ Onde fica o _chatGPT_ nessa história?
- __1.5__ Quais áreas da Ciência fazem parte do desenvolvimento de AM?

<font size=3>

__Respostas...__

- __1.1:__ AM trata-se de uma subárea de IA, onde modelos se utilizam de um conjunto de dados para "aprender" e realizar novas tarefas.
- __1.2:__ "Aprendizado" aqui é a capacidade de modelos identificar padrões nos dados, armazenar "conhecimento" e reproduzir respostas baseado no que foi treinado/aprendido.
- __1.3:__ Criar _agentes inteligentes_ para reproduzir respostas, extrair informações de dados, executar tarefas dadas certas condições.
- __1.4:__ O chatGPT é um modelos robusto desenvolvido com abordagens de redes neurais (Aprendizado Profundo), a qual é subárea de AM.

<center>
<img src="https://drive.google.com/uc?id=1Vt2NcbQf-CGB5IcSyiKmoE8Mbfeksk3O" width=300/>
</center>
<center>
<em> AI - Artificial Inteligence; ML - Machine Learning; DL - Deep Learning</em>
</center>

- __1.5:__ Probabilidade e Estatística, Ciências Biológicas, Teoria da Informação; além de Matemática e Ciência da Computação.

<br/><br/>

__Linha do tempo__ com os principais avanços na área de Aprendizado de Máquina:

- **1950s** – Alan Turing propõe o **Teste de Turing** para avaliar a inteligência artificial; primeiros modelos de redes neurais (Perceptron de Rosenblatt em 1958);
- **1960s-1970s** – Surgem os primeiros algoritmos de aprendizado estatístico (Sistemas Especialistas), como k-NN e regressão logística;
- **1980s** – Avanços em redes neurais com o algoritmo de retropropagação (Backpropagation), tornando o treinamento de redes multicamadas viável;
- **1990s** – Explosão dos métodos baseados em estatística e aprendizado supervisionado, como SVMs (Máquinas de Vetores de Suporte) e árvores de decisão;
- **2000s** – Crescimento do aprendizado de máquina com grandes volumes de dados e o surgimento de **Random Forests** e **Gradient Boosting**;
- **2010s** – Renascimento das **redes neurais profundas** (Deep Learning) com o uso massivo de GPUs e arquiteturas como CNNs para visão computacional e RNNs para processamento de linguagem natural;
- **2020s** – Domínio dos **modelos de linguagem de grande escala** (como GPT e BERT), avanços em **aprendizado por reforço** e IA generativa (como DALL·E e Stable Diffusion).

---


## __2. Aprendendo com dados:__
<font size=3>

Um modelo de AM se utiliza de um __conjunto de dados__ para treinamento e realiar tarefas. Estes dados podem ser do tipo _linguagem natural_, _áudio_, _imagens_, _vídeos_ e dados abstratos (_e.g._ dados clínicos e dados de vendas). Usualmente, esses dados podem formar uma tabela onde temos __amostras__ _vs_ __valores de característica/variáveis/atributos__.

<font size=3>

Vamos considerar como exemplo ilustrativo o conjunto de dados (clássico) de __flores Iris__.

<center>
<img src="https://drive.google.com/uc?id=13egHu4DxSm2rjfaGfyVIlEUf6vFVGWqt" width=250/>
</center>

- O conjunto de dados contém 150 amostras de flores Íris;
- Contém 50 amostras de cada uma das três espécies (_Iris setosa_, _Iris virginica_ e _Iris versicolor_);
- Cada amostra tem quatro características: comprimento e largura das sépalas e pétalas.


In [ ]:
import pandas as pd
from sklearn.datasets import load_iris

# Carregando o conjunto de dados:
iris = load_iris()

# Convertendo o dicionário "iris" em pandas.DataFrame:
df = pd.DataFrame(data=iris.data, columns=iris.feature_names)

# Adicionar a coluna que indica as espécies:
df['species'] = iris.target

# Transformar as etiquetas numéricas (0, 1, 2) em nos nomes das espécies:
df['species'] = df['species'].map({0: 'setosa', 1: 'versicolor', 2: 'virginica'})

# Apresentar as 5 primeiras linhas do dataframe:
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

  species  
0  setosa  
1  setosa  
2  setosa  
3  setosa  
4  setosa

## __3. Tarefas de aprendizado:__
<font size=3>

Acima, nós temos as __amostras__ de flores e seus __atributos__: comprimento da sépala, largura da sépala, comprimento da pétala, largura da pétala, espécie.

Em Aprendizado de Máquina existem três abordagens (principais) de aprendizagem:

<font size=3>
    
- __Aprendizado Supervisionado:__ o modelo realiza __predições__ de atributos. O conjunto de dados se apresenta na forma de atributos de entrada $x$ correlacionados a atributos de saída $y$. Assim, podemos obter uma nova predição $\bar y$ dado os novos atributos $\bar x$. Por exemplo, podemos considerar $x=(comp\_sépala,\, larg\_sépala,\, comp\_ pétala,\, larg\_pétala)$ e termos como variável álvo $y = (espécie)$ para realizarmos a tarefa de _classificação_.

- __Aprendizado Não-supervisionado:__ o modelo realiza a __descrição__ ou __exploração__ de atributos. O conjunto de dados se apresenta _apenas_ na forma de atributos de entrada $x$. Por exemplo, podemos considerar as métricas de pétala e sépala como $x$ a fim de realizarmos um agrupamento dos dados e verificar se os dados se separam por espécies.

<center>
<img src="https://drive.google.com/uc?id=171bo3MFVhvT_OAYbHr6y3fKev96qx9tm" width=300/>
</center>

- __Aprendizado por Reforço:__ o modelo realiza __ações__ baseado nos atributos de entrada a fim de _minimizar_ ou _maximizar_ a pontuação dada às ações. Por exemplo, no clássico jogo do Super Mario, onde temos as ações $x = ($⬅️ ⬆️ ➡️$)$, queremos saber quais movimentos $x$ maximizam a pontuação da fase.

<center>
<img src="https://drive.google.com/uc?id=1vtU5PJl8vAyuE8309sE2OgTh2ieN29J_" width=300/>
</center>

## __4. Indução de hipótese:__
<font size=3>

Para o treinamento de um modelo de AM, formulamos __hipóteses__ sobre quais atributos do conjunto de dados o modelo deve extrair para melhor representar as informações e gerar previsões precisas.




<font size=3>

- Hipóteses muito genéricas ou pouco informativas podem resultar em __subajuste__ (_underfitting_), tornando o modelo incapaz de capturar padrões relevantes e aprender adequadamente a relação entre entrada e saída.\
Por exemplo, nos dados de íris, caso deixemos de lado os atributos relacionados a sépalas, o modelo pode se tornar pouco preciso na classificação da espécie.

- Já hipóteses excessivamente específicas podem levar ao __superajuste__ (_overfitting_), onde o modelo memoriza padrões do treinamento, mas falha em generalizar para novos dados.\
Por exemplo, caso ajam 500 amostras de íris setosa, 400 amostras de íris versicolor e 20 amostras de íris virgínica, o modelo poderá apresentar um desempenho médio muito bom, mas uma performace ruim quando testado com atributos da espécie virgínica.


## __5. Viés indutivos:__
<font size=3>

O __viés indutivo__ é um conjunto de restrições que orientam o modelo durante o treinamento, ajudando a definir quais padrões ele deve priorizar.

<font size=3>

- Quais atributos das minhas amostras podem melhor levar uma classificação correta?
- Irei considerar todas as amostras?

No entanto, quando os dados de treino são enviesados &mdash; por exemplo, representando apenas um único gênero, raça ou classe social &mdash; o modelo pode aprender esses vieses e reproduzi-los nas previsões, comprometendo sua capacidade de generalização e gerando resultados discriminatórios, falsos ou imprecisos.


## __Referência:__
<font size=3>

- **(1.1 - 1.4)**: K. Faceli, Inteligência artificial: uma abordagem de aprendizado de máquina. Grupo Gen - LTC, 2011.

<br>

### __Material complementar:__
<font size=3>
    
 - [A história da Inteligência Artificial.](https://www.youtube.com/watch?v=Lhu8bdmkMCM)

   